README.md
# PostgreSQL database driver for Dart

[![Build Status](https://travis-ci.org/xxgreg/dart_postgresql.svg?branch=master)](https://travis-ci.org/xxgreg/dart_postgresql/)

## Basic usage

### Obtaining a connection

```dart
var uri = 'postgres://username:password@localhost:5432/database';
connect(uri).then((conn) {
	// ...
});
```

### SSL connections

Set the sslmode to require by appending this to the connection uri. This driver only supports sslmode=require, if sslmode is ommitted the driver will always connect without using SSL.

```dart
var uri = 'postgres://username:password@localhost:5432/database?sslmode=require';
connect(uri).then((conn) {
	// ...
});
```

### Querying

```dart
conn.query('select color from crayons').toList().then((rows) {
	for (var row in rows) {
		print(row.color); // Refer to columns by name,
		print(row[0]);    // Or by column index.
	}
});
```

### Executing

```dart
conn.execute("update crayons set color = 'pink'").then((rowsAffected) {
	print(rowsAffected);
});
```

### Query Parameters

Query parameters can be provided using a map. Strings will be escaped to prevent SQL injection vulnerabilities.

```dart
conn.query('select color from crayons where id = @id', {'id': 5})
  .toList()
	.then((result) { print(result); });

conn.execute('insert into crayons values (@id, @color)',
             {'id': 1, 'color': 'pink'})
	.then((_) { print('done.'); });
```

### Closing the connection

You must remember to call Connection.close() when you're done. This won't be
done automatically for you.

### Conversion of Postgresql datatypes.

Below is the mapping from Postgresql types to Dart types. All types which do not have an explicit mapping will be returned as a String in Postgresql's standard text format. This means that it is still possible to handle all types, as you can parse the string yourself.

```
     Postgresql type                 Dart type
	boolean                         bool
	int2, int4, int8                int
	float4, float8                  double
	numeric                         String
	timestamp, timestamptz, date    Datetime
	json, jsonb                     Map/List
	All other types                 String
```

### Mapping the results of a query to an object

```dart
class Crayon {
	String color;
	int length;
}

conn.query('select color, length from crayons')
	.map((row) => new Crayon()
	                     ..color = row.color
	                     ..length = row.length)
	.toList()
	.then((List<Crayon> crayons) {
		for (var c in crayons) {
			print(c is Crayon);
			print(c.color);
			print(c.length);
		}
	});
```

Or for an immutable object:

```dart
class ImmutableCrayon {
	ImmutableCrayon(this.color, this.length);
	final String color;
	final int length;
}

conn.query('select color, length from crayons')
  .map((row) => new ImmutableCrayon(row.color, row.length))
     .toList()
	.then((List<ImmutableCrayon> crayons) {
		for (var c in crayons) {
			print(c is ImmutableCrayon);
			print(c.color);
			print(c.length);
		}
	});
```

### Query queueing

Queries are queued and executed in the order in which they were queued.

So if you're not concerned about handling errors, you can write code like this:

```dart
conn.execute("create table crayons (color text, length int)");
conn.execute("insert into crayons values ('pink', 5)");
conn.query("select color from crayons").single.then((crayon) {
	print(crayon.color); // prints 'pink'
});
```

### Query streaming

Connection.query() returns a Stream of results. You can use each row as soon as
it is received, or you can wait till they all arrive by calling Stream.toList().

### Connection pooling

In server applications, a connection pool can be used to avoid the overhead of obtaining a connection for each request.

```dart
import 'package:postgresql/pool.dart';

main() {
  var uri = 'postgres://username:password@localhost:5432/database';
  var pool = new Pool(uri, minConnections: 2, maxConnections: 5);
  pool.messages.listen(print);
  pool.start().then((_) {
    print('Min connections established.');
    pool.connect().then((conn) { // Obtain connection from pool
      conn.query("select 'oi';")
        .toList()
        .then(print)
        .then((_) => conn.close()) // Return connection to pool
        .catchError((err) => print('Query error: $err'));
    });
  });
}
```

### Example program

Add postgresql to your pubspec.yaml file, and run pub install.

```
name: postgresql_example
dependencies:
  postgresql: any
```

```dart
import 'package:postgresql/postgresql.dart';

void main() {
  var uri = 'postgres://testdb:password@localhost:5432/testdb';
  var sql = "select 'oi'"; 
  connect(uri).then((conn) {
    conn.query(sql).toList()
    	.then((result) {
    		print('result: $result');
    	})
    	.whenComplete(() {
    		conn.close();
    	});
  });
}
```

## Testing

To run the unit tests you will need to create a database, and edit
'test/config.yaml' accordingly.

### Creating a database for testing

Change to the postgres user and run the administration commands.
```bash
sudo su postgres
createuser --pwprompt testdb
  Enter password for new role: password
  Enter it again: password
  Shall the new role be a superuser? (y/n) n
  Shall the new role be allowed to create databases? (y/n) n
  Shall the new role be allowed to create more new roles? (y/n) n
createdb --owner testdb testdb
exit
```

Check that it worked by logging in.
```bash
psql -h localhost -U testdb -W
```

Enter "\q" to quit from the psql console.

## License

BSD

## Links

http://www.postgresql.org/docs/9.2/static/index.html
http://www.dartlang.org/


CHANGELOG.md
### Version 0.3.4
 
 * Update broken crypto dependency.

#### Version 0.3.3

 * Fix #73 Properly encode/decode connection uris. Thanks to Martin Manev.
 * Permit connection without a password. Thanks to Jirka Daněk.

#### Version 0.3.2

 * Improve handing of datetimes. Thanks to Joe Conway.
 * Remove manually cps transformed async code.
 * Fix #58: Establish connections concurrently. Thanks to Tom Yeh.
 * Fix #67: URI encode db name so spaces can be used in db name. Thanks to Chad Schwendiman.
 * Fix #69: Empty connection pool not establishing connections.

#### Version 0.3.1+1

 * Expose column information via row.getColumns(). Credit to Jesper Håkansson for this change.

#### Version 0.3.0

  * A new connection pool with more configuration options.
  * Support for json and timestamptz types.
  * Utc time zone support.
  * User customisable type conversions.
  * Improved error handling.
  * Connection.onClosed has been removed.
  * Some api has been renamed, the original names are still functional but marked as deprecated.
      * import 'package:postgresql/postgresql_pool.dart'  =>  import 'package:postgresql/pool.dart'
      * Pool.destroy() => Pool.stop()
      * The constants were upper case and int type. Now typed and lower camel case to match the style guide.
      * Connection.unhandled => Connection.messages
      * Connection.transactionStatus => Connection.transactionState

  Thanks to Tom Yeh and Petar Sabev for their helpful feedback.


lib/settings.dart
part of postgresql;

final String DEFAULT_HOST = "localhost";
final num DEFAULT_PORT = 5432;

/**
 * Settings for PostgreSQL.
 */
class Settings {
  String _host;
  int _port;
  String _user;
  String _password;
  String _database;
  bool _requireSsl;
  
  /**
   * Settings map keys
   */
  static final String HOST = "host";
  static final String PORT = "port";
  static final String USER = "user";
  static final String PASSWORD = "password";
  static final String DATABASE = "database";

  Settings(this._host,
      this._port,
      this._user,
      this._password,
      this._database,
      {bool requireSsl: false})
    : _requireSsl = requireSsl;

  factory Settings.fromUri(String uri) {

    var u = Uri.parse(uri);
    if (u.scheme != 'postgres' && u.scheme != 'postgresql')
      throw new FormatException('Invalid uri.');

    if (u.userInfo == null || !u.userInfo.contains(':'))
      throw new FormatException('Invalid uri.');

    var userInfo = u.userInfo.split(':');

    if (u.path == null || !u.path.startsWith('/'))
      throw new FormatException('Invalid uri.');

    bool requireSsl = false;
    if (u.query != null)
      requireSsl = u.query.contains('sslmode=require');

    return new Settings(
        u.host,
        u.port == null ? DEFAULT_PORT : u.port,
        userInfo[0],
        userInfo[1],
        u.path.substring(1, u.path.length), // Remove preceding forward slash.
        requireSsl: requireSsl);
  }

  /**
   * Parse a map, apply default rules etc.
   * 
   * Throws [FormatException] when a setting (without default value)
   * is present.
   */
  Settings.fromMap(Map config){
    final String host = config.containsKey(HOST) ?
        config[HOST] : DEFAULT_HOST;
    final int port = config.containsKey(PORT) ?
        config[PORT] is int ? config[PORT]
          : throw new FormatException("Specified port is not a valid number")
        : DEFAULT_PORT;
    if (!config.containsKey(USER))
      throw new FormatException(USER);
    if (!config.containsKey(PASSWORD))
      throw new FormatException(PASSWORD);
    if (!config.containsKey(DATABASE))
      throw new FormatException(DATABASE);
    
    this._host = config[HOST];
    this._port = port;
    this._user = config[USER];
    this._password = config[PASSWORD];
    this._database = config[DATABASE];

    this._requireSsl = config.containsKey('sslmode') 
        && config['sslmode'] == 'require';
  }
  
  String get host => _host;
  int get port => _port;
  String get user => _user;
  String get password => _password;
  String get database => _database;
  bool get requireSsl => _requireSsl;

  /**
   * Return connection URI.
   * 
   * TODO
   * Support http://www.postgresql.org/docs/9.2/static/libpq-connect.html#AEN38149
   */
  String toUri()
    => "postgres://$_user:$_password@$_host:$_port/$_database${requireSsl ? '?sslmode=require' : ''}";
  
  String toString()
    => "Settings: [host: $_host, port: $_port, user: $_user, database: $_database]";

  Map toMap() {
    var map = new Map<String, dynamic>();
    map[HOST] = host;
    map[PORT] = port;
    map[USER] = user;
    map[PASSWORD] = password;
    map[DATABASE] = database;
    if (requireSsl)
      map['sslmode'] = 'require';
    return map;
  }
}